pkg/store/proxy_heap.go

```diff
@@ -256,12 +256,92 @@ func (h *ProxyResponseHeap) At() *storepb.SeriesResponse {
 	return atResp
 }
 
-func (l *lazyRespSet) StoreID() string {
-	return l.st.String()
+// dedupReadyResponse contains the original response with replica labels at the end of the labelset.
+// Labels used for query pushdown are placed right before deduplication labels.
+// dedupReadyResponse contains a signature hash calculated from all non-replica labels.
+type dedupReadyResponse struct {
+	*storepb.SeriesResponse
+
+	signature         uint64
+	hasPushdownMarker bool
+	nonReplicaLabels  labels.Labels
 }
 
-func (l *lazyRespSet) Labelset() string {
-	return labelpb.PromLabelSetsToString(l.st.LabelSets())
+func newDedupReadyResponse(resp *storepb.SeriesResponse, replicaLabels []string, replicaLabelSet map[string]struct{}) *dedupReadyResponse {
+	s := resp.GetSeries()
+	if s == nil {
+		return newUnsignedResponse(resp)
+	}
+
+	// TODO(fpetkovski): Push this computation to stores.
+	hasPushdownMarker := false
+	promLabels := labelpb.ZLabelsToPromLabels(resp.GetSeries().Labels)
+	if len(replicaLabels) == 0 {
+		for _, l := range promLabels {
+			if l.Name == dedup.PushdownMarker.Name {
+				hasPushdownMarker = true
+				break
+			}
+		}
+		return &dedupReadyResponse{
+			signature:         promLabels.Hash(),
+			nonReplicaLabels:  promLabels,
+			hasPushdownMarker: hasPushdownMarker,
+			SeriesResponse:    resp,
+		}
+	}
+
+	sort.Slice(s.Labels, func(i, j int) bool {
+		if _, ok := replicaLabelSet[s.Labels[i].Name]; ok {
+			return false
+		}
+		if _, ok := replicaLabelSet[s.Labels[j].Name]; ok {
+			return true
+		}
+		// Ensure that dedup marker goes just right before the replica labels.
+		if s.Labels[i].Name == dedup.PushdownMarker.Name {
+			hasPushdownMarker = true
+			return false
+		}
+		if s.Labels[j].Name == dedup.PushdownMarker.Name {
+			hasPushdownMarker = true
+			return true
+		}
+		return s.Labels[i].Name < s.Labels[j].Name
+	})
+
+	promLabels = labels.NewBuilder(promLabels).Del(replicaLabels...).Labels()
+	return &dedupReadyResponse{
+		signature:         promLabels.Hash(),
+		nonReplicaLabels:  promLabels,
+		hasPushdownMarker: hasPushdownMarker,
+		SeriesResponse:    resp,
+	}
+}
+
+func newUnsignedResponse(series *storepb.SeriesResponse) *dedupReadyResponse {
+	return &dedupReadyResponse{
+		signature:         0,
+		hasPushdownMarker: false,
+		SeriesResponse:    series,
+	}
+}
+
+func (m *dedupReadyResponse) LessWithoutReplicaLabels(other *dedupReadyResponse) bool {
+	if m.GetSeries() != nil && other.GetSeries() != nil {
```

Comment:
It crashes for me here. Generated code looks like this:

```go
	if x, ok := m.GetResult().(*SeriesResponse_Series); ok {
		return x.Series
	}
```

It doesn't check whether `x` is not `nil` :'/ Maybe we can add a small helper:

```go

func (m *SeriesResponse) GetSeriesNotNil() *Series {
	if x, ok := m.GetResult().(*SeriesResponse_Series); ok {
		if x == nil {
			return nil
		}
		return x.Series
	}
	return nil
}
```

The problem might be somewhere else, though, that we are comparing `nil` with not `nil`. Maybe related to https://github.com/thanos-io/thanos/pull/5717?

See my comment there.

Reply:
Thanks for testing this. I assume you are using the lazy strategy when it crashes?